### PR TITLE
Adding FastCAM Functionality

### PR DESCRIPTION
**Pull request by ryanchankh**
Hi Captum Developers,

We want to add FastCAM, a attribution method that uses information at the end of each network scale which is then  combined into a single saliency map, to the current Captum repository. We implemented our method in a file named `multiscale_fast_cam.py`, and created tests cases for our method. I have attached the links to the method, including a jupyter notebook demo, below. 

Links: 
[paper](https://arxiv.org/abs/1911.11293) 
[original repo](https://github.com/LLNL/fastcam)
[jupyter notebook demo](https://github.com/LLNL/fastcam/blob/captum/demo-captum.ipynb)

### CODE REVIEW

**Comment by vivekmig** [comment]
Hi @ryanchankh, sorry for the delay in getting back to you on this PR. We are currently in the process of determining our contribution guidelines and developing a consistent criteria for adding new methods, since we are responsible for maintaining all contributed methods in the future and supporting issues / bugs, etc. We will be releasing those guidelines soon, but since we hadn't released the criteria before your PR, we would be happy to accept FastCAM with some changes to ensure consistency with other methods and more flexibility for variants. We may document FastCAM as a new / experimental method along with other recently published methods in order to ensure new users aren't overwhelmed with too many choices.

Alternatively, if you don't have time to make these changes, we can also add your fork with FastCAM to the Awesome List, we'd be happy to merge that sooner to share your great work! Let us know which option you prefer. 

If you would like to continue with updating this diff to merge into the main Captum repo, we will do a more thorough code review and provide some suggestions. It would also be really helpful if you can provide some example outputs with your implementation and potentially reproduce some results from the paper, that would be helpful when reviewing the implementation to ensure correctness.

Thanks again for creating this, we really appreciate the contribution!

**Comment by ryanchankh** [comment]
No worries about the delay! I am happy to keep updating my current implementation with the newest codebase you have at `master`. For instance, I see that you just added Multi-layer support, which is exactly something that I can use, and I am happy to update that within this week. 

I can also add the FastCAM to the Awesome List for now. The implementation for FastCAM is actually fairly simple and straightforward. All the functions you will need are already implemented, and should be easy to update if your guidelines do change. 

The results in the paper are already reproduced using the implemented I had. You can see it in the demo here. I admit that test cases can be updated in a more thorough manner. I am happy to wait for your feedback and suggestions. 

Thank you for getting back to me! For now, let me add FastCAM to the awesome list and update my code with the multi-layer support.

**Comment by vivekmig** [comment]
> No worries about the delay! I am happy to keep updating my current implementation with the newest codebase you have at `master`. For instance, I see that you just added Multi-layer support, which is exactly something that I can use, and I am happy to update that within this week.
> 
> I can also add the FastCAM to the Awesome List for now. The implementation for FastCAM is actually fairly simple and straightforward. All the functions you will need are already implemented, and should be easy to update if your guidelines do change.
> 
> The results in the paper are already reproduced using the implemented I had. You can see it in the demo here. I admit that test cases can be updated in a more thorough manner. I am happy to wait for your feedback and suggestions.
> 
> Thank you for getting back to me! For now, let me add FastCAM to the awesome list and update my code with the multi-layer support.

Awesome, sounds great! I have merged my multi-layer support diff, so you should be able to rebase with that now. We'll also take a closer look next week and provide more detailed code review comments. 

In the meantime, the addition to the AwesomeList looks good! Just have a minor style change request there, can merge that soon. Thanks again for the great contribution 👍 !

**Comment by ryanchankh** [comment]
I have updated the code such that it uses the multi-layer `LayerActivation()` code. What other updates should I add to my code to match the contribution guidelines? 

Thanks~

**Comment by vivekmig** [comment]
> I have updated the code such that it uses the multi-layer `LayerActivation()` code. What other updates should I add to my code to match the contribution guidelines?
> 
> Thanks~

Thanks for updating with LayerActivation @ryanchankh ! I am finishing a review and will share comments soon, by the end of the week.

File: tests/attr/test_multiscale_fastcam.py [hunk 1]
```diff
@@ -0,0 +1,2024 @@
...
+import unittest
+from typing import Any, Dict
+
+import torch
+
+from captum.attr._core.multiscale_fast_cam import MultiscaleFastCam
+
+from ..helpers.basic import BaseTest, assertTensorAlmostEqual
+from ..helpers.basic_models import BasicModel_ConvNet
+
```
Comment (vivekmig) [inline tests/attr/test_multiscale_fastcam.py:11]: It would also be good to add DataParallel tests and ensure all functionality works appropriately on GPU. You can take a look at the test generator here: https://github.com/pytorch/captum/blob/master/tests/attr/test_data_parallel.py, which is based on this config https://github.com/pytorch/captum/blob/master/tests/attr/helpers/test_config.py.

**Review by vivekmig — COMMENTED** [review]
Sorry for the delay in getting back with feedback! Thanks again for adding this to Captum, it looks good overall @ryanchankh . I've added some comments on the API and tests, let us know if you have any questions.

**Comment by ryanchankh** [comment]
Hey @vivekmig, thank you for the feedback! And don't worry about the delays. Once I have everything fixed up and added, I'll let you know again in this thread.

**Comment by vivekmig** [comment]
Hi @ryanchankh , just wanted to follow up on this, will you still be able to make the updates to this PR? Thanks! cc: @NarineK

**Comment by ryanchankh** [comment]
Sorry I got caught up with some of my other projects this past month. I am aiming to finish updating this pull request by the end of this month. Is this timeline okay?

**Comment by vivekmig** [comment]
> Sorry I got caught up with some of my other projects this past month. I am aiming to finish updating this pull request by the end of this month. Is this timeline okay?

No problem, sure, that sounds great, thanks!

**Comment by ryanchankh** [comment]
Hi @vivekmig, I recently updated the `captum/attr/_core/multiscale_fast_cam.py` file based on your suggestions from above. In short, these are the changes:

1. I added the Gaussian normalization method, and made it the default choice for nomalization.
2. I added the mean, std, and normal entropy method scaling methods. This is to address the suggestion of having a comparison between different scales versus SMOE scale.
3. I also modified the code such that now the output of attribute matches with the same of the input. Essentially, there is an argument called `combine` (Defaults to True) that tells the method to combine the different scales. If the user want to get the individual scaled maps, one can simple set `combine=False`. Hopefully this better align with the code philosophy for this repository. 
4. Since the method has changed quite a bit, I also modified the test cases, and use smaller inputs for better readability. These outputs are based on the original repo [(link)](https://github.com/LLNL/fastcam), which are implemented by the original author Nathan. The test cases are quite simple, so please feel free to suggest if there should be any more comprehensive test cases. Most test cases I created references the ones from other attribution methods such as GradCAM. 
5. I also updated the doc string for the method. Since there are more arguments, please kindly look over them to see if they are clear to the user. 

As to the copyright issue, I am actually an employee at LLNL, working with the original authors. So there shouldn't be any copyright issues from the LLNL side. But if there is anything we need to do to ensure everything is okay legal-wise, I am willing to add any changes. 

Last but not least, I am having a hard time passing some of the automatic tests. It seems like it has something to do with the files I didn't touch. What is the best way to resolve these issues? 

Sorry for the long wait! And we really appreciate the feedback! Thank you!

File: captum/attr/_core/multiscale_fast_cam.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+                w = F.interpolate(
+                    smap[i].unsqueeze(0).unsqueeze(0),
+                    size=(height, width),
+                    mode=resize_mode,
+                    align_corners=False,
+                ).squeeze()
+                weighted_maps[i].append(w)
+                combined_map[i] += w * weights[m]
+        combined_map = combined_map / np.sum(weights)
+        weighted_maps = torch.stack([torch.stack(wmaps) for wmaps in weighted_maps])
```
Comment (vivekmig) [inline captum/attr/_core/multiscale_fast_cam.py:215]: It looks like this will now be shaped as num_examples x num_layers x 1 x height x width, is that correct? We should at least document this in the return type, since this would have one extra dimension than the input. Alternatively, the extra dimension of one could be avoided, or we could return a list of num_layers with each layer separately analogous to the layer methods.

**Review by vivekmig — COMMENTED** [review]
Thanks for addressing the comments @ryanchankh , this looks great :) ! I've added a few more minor comments. @NarineK may also take another look to provide some additional feedback. 

Regarding the CircleCI tests, I think the errors are related to issues we have fixed previously, can you try pulling the latest master into your branch? That should likely resolve the issues.

**Comment by krebso** [comment]
Hello there, any updates on this one?

**Comment by krebso** [comment]
@vivekmig is there a possibility you will merge this? If there is some work that needs to be done, LMK, I am more than willing to help. Thanks
